ssl/ssl_lib.c

```diff
@@ -4344,3 +4344,118 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+                                    const uint8_t *encrypted_premaster,
+                                    size_t encrypted_premaster_len,
+                                    const uint8_t *premaster,
+                                    size_t premaster_len)
+{
+    char *out = NULL;
+    char *cursor = NULL;
+    size_t out_len = 0;
+    size_t i;
+
+    if (ssl->ctx->keylog_callback == NULL) return 1;
+
+
+    if (encrypted_premaster_len < 8) {
+        SSLerr(SSL_F_SSL_LOG_RSA_CLIENT_KEY_EXCHANGE, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    /*
+     * Our output buffer will contain:
+     * - The characters "RSA " (4 octets)
+     * - The first 8 bytes of the encrypted pre-master secret, rendered in
+     *   ASCII hexadecimal. (16 octets)
+     * - A space. (1 octet)
+     * - The pre-master secret, rendered in ASCII hexadecimal. (2 * the length
+     *   of the pre-master secret octets)
+     * - A trailing NULL.
+     */
+    out_len = 4 + 16 + 1 + premaster_len*2 + 1;
+    if ((out = cursor = OPENSSL_malloc(out_len)) != NULL) {
+        return 0;
+    }
+
+    strcpy(cursor, "RSA ");
+    cursor += 4;
+
+    for (i = 0; i < 8; i++) {
+        sprintf(cursor, "%02x", encrypted_premaster[i]);
+        cursor += 2;
+    }
+    *cursor++ = ' ';
+
+    for (i = 0; i < premaster_len; i++) {
+        sprintf(cursor, "%02x", premaster[i]);
+        cursor += 2;
+    }
+    *cursor = '\0';
+
+    ssl->ctx->keylog_callback(ssl, (const char *)out);
+    OPENSSL_free(out);
+    return 1;
+}
+
+int ssl_log_master_secret(SSL *ssl,
+                          const uint8_t *client_random,
+                          size_t client_random_len,
+                          const uint8_t *master,
+                          size_t master_len)
+{
+    char *out = NULL;
+    char *cursor = NULL;
+    size_t out_len = 0;
+    size_t i;
+
+    if (ssl->ctx->keylog_callback == NULL) return 1;
+
+    if (client_random_len != 32) {
+        SSLerr(SSL_F_SSL_LOG_MASTER_SECRET, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    /*
+     * Our output buffer will contain:
+     * - The characters "CLIENT_RANDOM " (14 octets)
+     * - The 32 bytes of the client random, rendered in ASCII hexadecimal. (64
+     *   octets)
+     * - A space. (1 octet)
+     * - The master secret, rendered in ASCII hexadecimal. (2 * the length of the
+     *   master secret octets)
+     * - A trailing NULL.
+     */
+    out_len = 14 + 64 + 1 + master_len*2 + 1;
+    if (!(out = cursor = OPENSSL_malloc(out_len))) {
+        return 0;
+    }
+
+    strcpy(cursor, "CLIENT_RANDOM ");
+    cursor += 14;
+    for (i = 0; i < 32; i++) {
+        sprintf(cursor, "%02x", client_random[i]);
+        cursor += 2;
+    }
+    *cursor++ = ' ';
+
+    for (i = 0; i < master_len; i++) {
+        sprintf(cursor, "%02x", master[i]);
+        cursor += 2;
+    }
+    *cursor = '\0';
+
+    ssl->ctx->keylog_callback(ssl, (const char *)out);
+    OPENSSL_free(out);
+    return 1;
+}
```

Comment:
These two functions have a *lot* in common.  Would it be possible to do some code reuse rather than code copying, i.e. have a common function that takes some extra parameters, say:

``` C
static int nss_keylog_int(const char *prefix,
                          SSL *ssl,
                          const uint8_t *p1, size_t p1_len,
                          const uint8_t *p2, size_t p2_len)
```

And then have call it like this?

``` C
int ssl_log_rsa_client_key_exchange(SSL *ssl,
                                    const uint8_t *encrypted_premaster,
                                    size_t encrypted_premaster_len,
                                    const uint8_t *premaster,
                                    size_t premaster_len)
{
    if (encrypted_premaster_len < 8) {
        SSLerr(SSL_F_SSL_LOG_RSA_CLIENT_KEY_EXCHANGE, ERR_R_INTERNAL_ERROR);
        return 0;
    }
    return nss_keylog_int("RSA",
                          ssl,
                          encrypted_premaster, 8,
                          premaster, premaster_len);
}

int ssl_log_master_secret(SSL *ssl,
                          const uint8_t *client_random,
                          size_t client_random_len,
                          const uint8_t *master,
                          size_t master_len)
{
    if (client_random_len != 32) {
        SSLerr(SSL_F_SSL_LOG_MASTER_SECRET, ERR_R_INTERNAL_ERROR);
        return 0;
    }
    return nss_keylog_int("CLIENT_RANDOM",
                          ssl,
                          client_random, 32,
                          master, master_len);
}

```